lib/Conversion/FIRRTLToHW/LowerToHW.cpp

```diff
@@ -1650,42 +1649,78 @@ Value FIRRTLLowering::getReadInOutOp(Value v) {
   return result;
 }
 
-void FIRRTLLowering::addToAlwaysBlock(Value clock,
-                                      std::function<void(void)> fn) {
-  auto &op = alwaysBlocks[{clock, builder.getBlock()->getParent()}];
-  if (op) {
-    runWithInsertionPointAtEndOfBlock(fn, op.body());
-    // Move the earlier always block(s) down to where the last would have been
-    // inserted.  This ensures that any values used by the always blocks are
-    // defined ahead of the uses, which leads to better generated Verilog.
-    op->moveBefore(builder.getInsertionBlock(), builder.getInsertionPoint());
-  } else {
-    op = builder.create<sv::AlwaysOp>(sv::EventControl::AtPosEdge, clock, fn);
-  }
-}
-
-void FIRRTLLowering::addToAlwaysFFBlock(sv::EventControl clockEdge, Value clock,
-                                        ::ResetType resetStyle,
-                                        sv::EventControl resetEdge, Value reset,
-                                        std::function<void(void)> body,
-                                        std::function<void(void)> resetBody) {
-  auto &op = alwaysFFBlocks[std::make_tuple(
-      builder.getBlock(), clockEdge, clock, resetStyle, resetEdge, reset)];
-  if (op) {
-    runWithInsertionPointAtEndOfBlock(body, op.bodyBlk());
-    runWithInsertionPointAtEndOfBlock(resetBody, op.resetBlk());
+void FIRRTLLowering::addToAlwaysBlock(sv::EventControl clockEdge, Value clock,
+                                      ::ResetType resetStyle,
+                                      sv::EventControl resetEdge, Value reset,
+                                      std::function<void(void)> body,
+                                      std::function<void(void)> resetBody) {
+  auto &op = alwaysBlocks[std::make_tuple(builder.getBlock(), clockEdge, clock,
+                                          resetStyle, resetEdge, reset)];
+  auto &alwaysOp = op.first;
+  auto &insideIfOp = op.second;
+  if (alwaysOp) {
+    if (reset) {
+      assert(insideIfOp && "reset body must be initialized before");
+      runWithInsertionPointAtEndOfBlock(resetBody, insideIfOp.thenRegion());
+      runWithInsertionPointAtEndOfBlock(body, insideIfOp.elseRegion());
+    } else {
+      runWithInsertionPointAtEndOfBlock(body, alwaysOp.body());
+    }
 
     // Move the earlier always block(s) down to where the last would have been
     // inserted.  This ensures that any values used by the always blocks are
     // defined ahead of the uses, which leads to better generated Verilog.
-    op->moveBefore(builder.getInsertionBlock(), builder.getInsertionPoint());
+    alwaysOp->moveBefore(builder.getInsertionBlock(),
+                         builder.getInsertionPoint());
   } else {
```

Comment:
Is there a way to rework this to share the code for the always block always exists.  E.g., can this do something like:

```c++
if (!alwaysOp) {
  /* Generate empty always block. */
  alwaysOp = /* ... */;
  insideIfOp = /* ... */;
}

if (reset) {
  runWithInsertionPointAtEndOfBlock(resetBody, /* ... */);
  runWithInsertionPointAtEndOfBlock(body, /* ... */);
} else {
  runWithInsertionPointAtEndOfBlock(body, /* ... */);
}
```